src/libsyntax/codemap.rs

```diff
@@ -276,97 +290,76 @@ pub fn original_sp(cm: &CodeMap, sp: Span, enclosing_sp: Span) -> Span {
-        Span { lo: lo, hi: hi, expn_id: expn_id }
+    pub fn from_span(primary_span: Span) -> MultiSpan {
+        MultiSpan {
+            primary_spans: vec![primary_span],
+            span_labels: vec![]
+        }
     }
 
-    /// Merges or inserts the given span into itself.
-    pub fn push_merge(&mut self, mut sp: Span) {
-        let mut idx_merged = None;
-
-        for idx in 0.. {
-            let cur = match self.spans.get(idx) {
-                Some(s) => *s,
-                None => break,
-            };
-            // Try to merge with a contained Span
-            if let Some(union) = cur.merge(sp) {
-                self.spans[idx] = union;
-                sp = union;
-                idx_merged = Some(idx);
-                break;
-            }
-            // Or insert into the first sorted position
-            if sp.hi <= cur.lo {
-                self.spans.insert(idx, sp);
-                idx_merged = Some(idx);
-                break;
-            }
-        }
-        if let Some(idx) = idx_merged {
-            // Merge with spans trailing the insertion/merging position
-            while (idx + 1) < self.spans.len() {
-                if let Some(union) = self.spans[idx + 1].merge(sp) {
-                    self.spans[idx] = union;
-                    self.spans.remove(idx + 1);
-                } else {
-                    break;
-                }
-            }
-        } else {
-            self.spans.push(sp);
+    pub fn from_spans(vec: Vec<Span>) -> MultiSpan {
+        MultiSpan {
+            primary_spans: vec,
+            span_labels: vec![]
         }
     }
 
-    /// Inserts the given span into itself, for use with `end_highlight_lines`.
-    pub fn push_trim(&mut self, mut sp: Span) {
-        let mut prev = mk_sp(BytePos(0), BytePos(0));
+    pub fn push_primary_span(&mut self, span: Span) {
+        self.primary_spans.push(span);
+    }
 
-        if let Some(first) = self.spans.get_mut(0) {
-            if first.lo > sp.lo {
-                // Prevent us here from spanning fewer lines
-                // because of trimming the start of the span
-                // (this should not be visible, because this method ought
-                // to not be used in conjunction with `highlight_lines`)
-                first.lo = sp.lo;
-            }
+    pub fn push_span_label(&mut self, span: Span, label: String) {
+        self.span_labels.push((span, label));
+    }
+
+    /// Selects the first primary span (if any)
+    pub fn primary_span(&self) -> Option<Span> {
+        self.primary_spans.first().cloned()
+    }
+
+    /// Returns all primary spans.
+    pub fn primary_spans(&self) -> &[Span] {
+        &self.primary_spans
+    }
+
+    /// Returns the strings to highlight. If we have an explicit set,
+    /// return those, otherwise just give back an (unlabeled) version
+    /// of the primary span.
+    pub fn span_labels(&self) -> Vec<SpanLabel> {
```

Comment:
@nrc they may or may not be in the list of labeled spans, that's the point. If they are not, we want to add them with no label.